nextstrain/cli/runner/aws_batch/jobs.py

```diff
@@ -219,16 +226,116 @@ def forwarded_environment() -> List[dict]:
+
+    This function only exists because the ``containerOverrides`` data structure
+    used in job submission doesn't support overriding the image.
+
+    Returns a job definition name.
+    """
+    # XXX TODO: We only create derived job definitions from an existing base
+    # definition, which is minimally modified.  This is a conservative choice
+    # that avoids hardcoding a default definition.  In the future, we might
+    # consider relaxing this to auto-create a default base definition.  The
+    # advantage there is one less piece of initial AWS setup.  The challenge is
+    # choosing defaults that work for everyone's taste for resources/spending,
+    # as well as using the right job roles and other details.
+    #   -trs, 22 May 2020
+    batch = aws.client_with_default_region("batch")
+
+    base_definition = lookup_definition(base_definition_name)
+
+    if not base_definition:
+        raise UserError("AWS Batch job definition «%s» does not exist" % base_definition_name)
+
+    # The base definition might already use this image, in which case we avoid
+    # creating anything.  This lets admins not grant job definition creation to
+    # users if they want; users will instead have to specify matching
+    # --aws-batch-job and --image values.
+    #
+    # Split the image name in order to normalize the implicit :latest tag.
+    base_image = base_definition["containerProperties"]["image"]
+
+    if split_image_name(base_image) == split_image_name(image):
+        return base_definition_name
+
+    # The revision of the base definition is included in the derived image name
+    # so that changes to the base definition cause derived definitions to be
+    # re-created.  Unfortunately, we can't explicitly set the revision of new
+    # definitions.
+    derived_definition_name = sanitize_name(
+        "_".join((
+            base_definition_name,
+            str(base_definition["revision"]),
+            *split_image_name(image))))
+
+    derived_definition = lookup_definition(derived_definition_name)
+
+    if not derived_definition:
+        derived_definition = deepcopy(base_definition)
+        derived_definition["jobDefinitionName"] = derived_definition_name
+        derived_definition["containerProperties"]["image"] = image
+
+        # These are AWS-assigned keys returned by describe_job_definitions() which
+        # aren't supported as keyword arguments by register_job_definition().
+        for key in {'jobDefinitionArn', 'revision', 'status'}:
+            del derived_definition[key]
+
+        batch.register_job_definition(**derived_definition)
+    else:
+        derived_image = derived_definition["containerProperties"]["image"]
+
+        assert split_image_name(derived_image) == split_image_name(image), (
+            "Expected job definition «%s», derived from «%s», to use image «%s», "
+            "but it uses «%s» instead"
+            % (derived_definition_name, base_definition_name, image, derived_image))
```

Comment:
Just out of curiosity, why use `%s` instead of `.format()`? (Looks like `%s` is used for string formatting for most of this repo)

Reply:
Astute observation! I find `.format()` slightly clunkier than printf-style most of the time, so tend not to use it. Just pure preference! I prefer f-strings the most, but those aren't supported until Python 3.6 and this project currently documents support for 3.5. Python 3.5 was still somewhat common on Linux computers when I was first starting this project in 2018. When this project moves to 3.6 or higher as a minimum Python version, I'd like to move to f-strings.